### ISSUE TITLE
Update ES6 module Imports for use with Webpack or Browserify

### ISSUE THREAD

**Issue by willcode4food** [photonstorm/lazer#8]

**Comment by willcode4food** [comment]
So I removed the branch (UMD_Webpack) I was working on to for UMD in favor of a new branch, Update_ES6_Imports_Relative.  I think the way to go is to leverage the existing work with ES6 modules and just make the imporst relative.  I'll also add index.js files to each sub directory and create tests to demonstrate the API for each library module.  We'll also be able to build using Webpack.

**Comment by willcode4food** [comment]
I completed updating all the lazer libraries with relative pathing in their ES6 import statements.  So now any bundler that supports ES6 modules through transpiling can work with the repo.  I use webpack regularly so I created a build step for use with it.  But one could easily use Browserify or Rollup or something else.    The built package is intended for usage similar to how lodash is imported into projects.  For example:

`   import { DegToRad } from 'lazer/math';
     import { Canvas, BackgroundColor, graphics } from 'lazer/canvas';
     import { AddToDOM } from 'lazer/dom';

     const { Line, Stroke, Rectangle } = graphics;`

This structure could easily be changed in the webpack.config.js by updating the entry points.  I tried to keep the commits to one per each sub library.  You can build lazer using webpack by updating to the branch [Update_ES6_Imports_Relative](https://github.com/photonstorm/lazer/tree/Update_ES6_Imports_Relative)

I'd love to get some feedback on this concept.

**Comment by photonstorm** [comment]
Hiya,

Wow - loads of work. Nice one!

I'm curious but what do the index.js files do? (like this one https://github.com/photonstorm/lazer/blob/3338a0ac7e712d9a390c257816be33cfc03a6480/src/time/index.js) some folders have them, but most don't.

What are the rules for when they're created, and how are they used?

Aside from that, this looks like a sweet move, so it makes a lot of sense to merge it all in sooner rather than later.

**Comment by willcode4food** [comment]
Thanks for the feedback.  So the index.js files are a way in ES6 to package up individually exported modules into a single discrete package.   This is helpful when importing into another module.  If you look at the tests for any of the libraries, such as [time](https://github.com/photonstorm/lazer/blob/Update_ES6_Imports_Relative/tests/time/time_spec.js)  you can see that we import the library by pointing to the sub-directory instead of the individual .js file.   You can also destructure to get individual modules.  For Example:

```javascript
import { Clock }  from '../../src/time'
```

This is a standard pattern when working with ES6 modules and bundling them up.  Another way I've seen this done is with individual package.json files with a main pointing to a root module that imports all the rest.  But I felt the index.js files were less invasive to the lazer repo.  

I felt that it made sense to keep the hierarchy already established by the directory structure that was in place.   So there actually should be one index.js file in each sub-directory.  Webpack wouldn't build properly if there wasn't.  So let me know if you see a specific example of one missing.  

The rules are simple, import the individual module unless a file exports multiple modules.  If multiple exports, import them using the following syntax

```javascript
import * as <package> syntax 
```

Or destructure as described above.  You don't need to include the .js extension in the import statements.  But I kept them in there to be less confusing to those not familiar with ES6.  I'd prefer not to have them along with semicolons, but that's a personal preference and I want to follow the coding standards of this team.  I think the hump for many new to ES6 is to think in terms of modules instead of in terms of individual scripts.  

The exports uses an ES6 object literal property shorthand to simplify something like this

```javascript
export {
     Clock: Clock 
}
```
to 

```javascript
export {
     Clock
}
```

I hope that clarifies things.  And a merge would be awesome.  I can also port over the SAT library once it is finished.  Let me know if you have any other questions or concerns.

**Comment by GGAlanSmithee** [comment]
@willcode4food 

I think you can simplify it further by doing this

```javascript
export * from './Clock';
export * from './MasterClock';

// or if you want to keep the named export 

export {* as Clock} from './Clock';
export {* as MasterClock} from './MasterClock';
```

Not 100% sure the syntax is correct, since I can't check ATM, but I think it is.
Also, notice I removed `.js` from the imports since that isn't needed by convention.

BTW. What does `'create index'` do? Never seen that before.

**Comment by photonstorm** [comment]
Hmm interesting. I'm not sure what to make of the `index.js` files. They sound far more sensible than one giant package.json (which is the wrong approach, as this is a monorepo, not a single package), but equally they bundle together things in the same folder which aren't necessarily meant to be used together.

For example the `create/palettes` folder has an index which includes all the available palettes so far. Say for arguments sake that folder grew to include a few hundred palettes, would the index be expected to track them all?

I assume if you then imported that index into your own project, but only used 1 palette, the rest would get shaken out during the build process anyway.

re: having the .js filename on the import lines - I'd rather not have it too, but SystemJS (which is used in the lazer-dev repo) fails miserably to load anything for me without it. If we can fix that, I'm happy to remove it.

**Comment by NTARelix** [comment]
The latest webpack should allow tree shaking to prevent these index files from importing unused modules as long as they use es6 imports and exports.

To me this looks like just one more thing to maintain when adding or removing modules for the benefit of minor convenience when it's just as easy to import a module directly.

My .02

**Comment by willcode4food** [comment]
Yes, I believe that syntax is valid and agreed much more simplified.  So the 'create index' snippet was created by a tool I used to generate as much of the code for the index.js files.  (There was quite a few).  That line can be pruned.  I tried catching them all, but a few slipped through

**Comment by willcode4food** [comment]
@photonstorm All great questions.  I understand the concerns around something like the ```create/palettes library.  And admittedly, I built the packaging strategy without really understanding any context about how the libraries would be used.

I think we should be having a larger discussion around how we want to package Lazer.  I've been thinking about this a lot and monorepo makes a lot of sense.  I really like the way lodash packages their enormous library.  You import what you want and they offer the ability for custom builds to address these types of issues.   There's a lot of options.   I looked at this [article](https://plot.ly/javascript/modularizing-monolithic-javascript-projects/) from the plotly guys for some direction.  

And regarding lazer-dev, I'm in the process of updating the JS in a similar fashion to support this new structure in Lazer.  You can try it [here](https://github.com/willcode4food/lazer-dev) on the webpack branch.  To run it use the following commands:

```bash
npm install 

npm run dev
```
Then point your browser to http://localhost:8080/index.html

With this type of app we won't need a dependency on SystemJS.  There are a lot of other benefits to building it as an SPA so let me know what you think of this approach.  (It's probably a whole different discussion.) 

@NTARelix Yes, tree shaking is supported by Webpack 2 (beta) and is already available in Rollup.  I believe this type of structure paves the way for optimizations.  Again, I think we ultimately need to decide how we want Lazer to be packaged.

**Comment by photonstorm** [comment]
To give some clarity to how I see the Lazer repo working:

1) It should be thought of as one giant single repository for all of the
Lazer building blocks.

2) There will absolutely be parts of the repo that contain multiple ways of
doing the same thing. For example we're likely to have a number of
renderers, a number of Transform handlers, etc. As the repo grows, so too
will the variety and scope of its contents.

3) it's perfectly OK if we end up with a number of functions (or modules)
that all do the same thing, as long as they go about it slightly
differently.

4) I don't want to version the repo as a whole. I know I'll have to, but I
don't ever want to be in the position of saying "wait for release X", I
want to say "this is fixed *right now*, go grab it". I think we will need
to version every function and module, to help devs who are cherry picking
parts, but Lazer won't follow the same release pattern as phaser does.

5) To help ease this I am considering creating other repos that use Lazer,
I.e. "Lazer Nano", that does follow the Phaser style release system, and
builds against the Lazer repo at a fixed point in time, so upstream changes
don't break older releases. These will be the repos we'll direct beginners
to.

I'll read over the article you linked, as I'm sure this isn't a unique
set-up. But it does need planning properly to make sure it doesn't go nuts.





On Monday, 16 May 2016, Marc Arbesman <notifications@github.com> wrote:

> @photonstorm <https://github.com/photonstorm> All great questions. I
> understand the concerns around something like the ```create/palettes
> library. And admittedly, I built the packaging strategy without really
> understanding any context about how the libraries would be used.
>
> I think we should be having a larger discussion around how we want to
> package Lazer. I've been thinking about this a lot and monorepo makes a lot
> of sense. I really like the way lodash packages their enormous library. You
> import what you want and they offer the ability for custom builds to
> address these types of issues. There's a lot of options. I looked at this
> article
> <https://plot.ly/javascript/modularizing-monolithic-javascript-projects/>
> from the plotly guys for some direction.
>
> And regarding lazer-dev, I'm in the process of updating the JS in a
> similar fashion to support this new structure in Lazer. You can try it
> here <https://github.com/willcode4food/lazer-dev> on the webpack branch.
> To run it use the following commands:
>
> npm install
>
> npm run dev
>
> Then point your browser to http://localhost:8080/index.html
>
> With this type of app we won't need a dependency on SystemJS. There are a
> lot of other benefits to building it as an SPA so let me know what you
> think of this approach. (It's probably a whole different discussion.)
>
> @NTARelix <https://github.com/NTARelix> Yes, tree shaking is supported by
> Webpack 2 (beta) and is already available in Rollup. I believe this type of
> structure paves the way for optimizations. Again, I think we ultimately
> need to decide how we want Lazer to be packaged.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/photonstorm/lazer/issues/8#issuecomment-219487134>
>


-- 
Photon Storm Ltd.
http://www.photonstorm.com

Skype: richard.davey
Twitter: @photonstorm

Registered in England and Wales.
Company ID: 8036404
VAT Number: 136 4333 27

**Comment by GGAlanSmithee** [comment]
@willcode4food @NTARelix 
AFAIK, as long as you use ES6 `import / export`and specify `esnext:main` in your `package.json` file, rollup and webpack 2 will be able to produce an optimized build. We had an interesting conversation regarding this in the `three.js` repo: mrdoob/three.js#7131 which is a library not to different from `Lazer` in some ways. Also, @rich-harris, the creator of `rollup` has an modularized version of `three.js` here: https://github.com/rollup/three-jsnext which can be used as a reference point.

@photonstorm 
As far as releases go, `node` and `npm` encourages the use of [semver](http://semver.org/).

In it's basic form it's easy to follow, from their homepage:

> 1. MAJOR version when you make incompatible API changes,
> 2. MINOR version when you add functionality in a backwards-compatible manner, and
> 3. PATCH version when you make backwards-compatible bug fixes.

If you want to do alpha or beta releases, you can do so using `tags`.
Just adding some info though. I have to admit I like seeing the names of your releases of `Phaser` :)

**Comment by photonstorm** [comment]
The problem with semver (in the GitHumb / npm sense) is that it applies to
the whole repo / package, not just to a specific part of it.

I was hoping to avoid ever having to say "This is Lazer release v1.0.4"
because Lazer itself will never be packaged in that way (as one single
whole entity). But rather I wanted to say "This is Lazer / Arcade Physics
v1.0.4" and "This is Lazer / Math / AngleBetween v1.2", etc. I.e. each
little part of it is versioned.

The problem of course is that npm just doesn't work like this from what I
can tell, and have seen. When you tag it, you tag the whole thing.

So now I'm thinking maybe I'll have to do "This is Lazer v1.0.4" and it
contains "AngleBetween v1.2, SpriteCollide v1.4, AudioMute v0.5" etc. Which
isn't ideal either, although I'm not sure of other choices?




Photon Storm Ltd.
http://www.photonstorm.com

Skype: richard.davey
Twitter: @photonstorm

Registered in England and Wales.
Company ID: 8036404
VAT Number: 136 4333 27

**Comment by GGAlanSmithee** [comment]
@photonstorm 

I see.

I do not know the exact reasons for chosing the monorepo approach, but just because code shares a `git` repo, it doesn't mean it has to be delivered as one `npm` package.

You could have one entry point (one `package.json`) per "module" and `npm publish` those modules under their respective names. Another "higher level" module would include them as actual dependencies (as in `npm install <module> --save`). For the enduser, this would act just as it was one package, since, `npm install lazer` would include all it's dependencies (be it "internal" or not), and a potential bundle would be bundled the same way. This could be handled nicely using a `npm namespace`.

I'm just thinking out loud though. I can see something like this adding to the maintenece overhead, and I'm not sure how much actual value it would add. I guess one benefit would be that it would make API surfaces and borders clear.

**Comment by photonstorm** [comment]
The main reason is maintainer friction (i.e. mine!), splitting it all up
into loads of tiny repos (one per module) would cause me a massive admin
headache, that is best avoided.

Yeah I'm aware of using multiple package.json in a single repo, and that
might be an option. It's still a bit of a headache having to publish each
one, each time it changes, but it provides proper distinctions between
releases.


Photon Storm Ltd.
http://www.photonstorm.com

Skype: richard.davey
Twitter: @photonstorm

Registered in England and Wales.
Company ID: 8036404
VAT Number: 136 4333 27

**Comment by willcode4food** [comment]
I am starting to understand the issue more now.  And in the article I linked to above, plotly specifically address admin issues as an argument for a monorepo.  One Bug tracker, point of contact, etc.  

@GGAlanSmithee my webpack config uses multiple entry points (one per library, i.e. components, canvas, time, renderer, etc.).  You can see the files in the root post-build step.  (/canvas.js, /time.js, etc.)  My thinking was so a developer when building a game with lazer can install via NPM and import like this

```javascript
import { Clock } from 'lazer/time'
```

So with this in mind maybe its a matter of one repo in github, and maybe we publish to NPM in multiple modules?  Just spit-ballin'

**Comment by photonstorm** [comment]
Yeah I'm beginning to conclude the same thing. A single repo with multiple
publish points. To be honest if it wasn't such an admin pain I would use
separate repos too, but that's a bad path I don't want to head down.

The only thing wrong with the index.js files as they stand is that they
assume one folder = one module, which isn't really the case. I completely
understand why you thought this of course.

Maybe the next best course of action is for me to devise a repo folder
structure and set some conventions in place?


On Tuesday, 17 May 2016, Marc Arbesman <notifications@github.com> wrote:

> I am starting to understand the issue more now. And in the article I
> linked to above, plotly specifically address admin issues as an argument
> for a monorepo. One Bug tracker, point of contact, etc.
>
> @GGAlanSmithee <https://github.com/GGAlanSmithee> my webpack config uses
> multiple entry points (one per library, i.e. components, canvas, time,
> renderer, etc.). You can see the files in the root post-build step.
> (/canvas.js, /time.js, etc.) My thinking was so a developer when building a
> game with lazer can install via NPM and import like this
>
> import { Clock } from 'lazer/time'
>
> So with this in mind maybe its a matter of one repo in github, and maybe
> we publish to NPM in multiple modules? Just spit-ballin'
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/photonstorm/lazer/issues/8#issuecomment-219757979>
>


-- 
Photon Storm Ltd.
http://www.photonstorm.com

Skype: richard.davey
Twitter: @photonstorm

Registered in England and Wales.
Company ID: 8036404
VAT Number: 136 4333 27

**Comment by NTARelix** [comment]
After a little research I think a structure similar to Babel might work well for a project like this: 
https://github.com/babel/babel/blob/master/doc/design/monorepo.md

They've made a tool to help with monorepo tasks
https://github.com/lerna/lerna

I'm going to look into this a little more later when I get home from work. Any thoughts on this?

**Comment by photonstorm** [comment]
Lerna looks fantastic! That will be super-useful, thank you. And yes, I
think this is very much like the Babel structure. Not exactly, but
certainly closer than anything else I've seen.


Photon Storm Ltd.
http://www.photonstorm.com

Skype: richard.davey
Twitter: @photonstorm

Registered in England and Wales.
Company ID: 8036404
VAT Number: 136 4333 27

**Comment by willcode4food** [comment]
Nice find!  I'll play around with this as well

**Comment by GGAlanSmithee** [comment]
+1 on lerna.. That must have been the tool used to [deploy babel 6 at embercamp](https://www.youtube.com/watch?v=Q_ncaTYEizc&feature=youtu.be&t=970)

BTW, just a notice, relying on folder structures for imports and communicating / encouraging that in documentation, means moving files / renaming folders etc becomes a breaking API change.

Just a reaction to:

> `import { Clock } from 'lazer/time'`

**Comment by willcode4food** [comment]
I think deciding on the intended module structure would be helpful moving forward.  The [Babel Package Structure](https://github.com/babel/babel/tree/master/packages)  is interesting.

**Comment by photonstorm** [comment]
I agree re: the folder moving or renaming issue. But what other option is
there? At some point, something has to map to a folder, so I'm not sure it
matters where it's set? : if it changes, it changes.




On Tuesday, 17 May 2016, Alan Smithee <notifications@github.com> wrote:

> +1 on lerna.. That must have been the tool used to deploy babel 6 at
> embercamp
> <https://www.youtube.com/watch?v=Q_ncaTYEizc&feature=youtu.be&t=970>
>
> BTW, just a notice, relying on folder structures for imports and
> communicating / encouraging that in documentation, means moving files /
> renaming folders etc becomes a breaking API change.
>
> Just a reaction to:
>
> import { Clock } from 'lazer/time'
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/photonstorm/lazer/issues/8#issuecomment-219878023>
>


-- 
Photon Storm Ltd.
http://www.photonstorm.com

Skype: richard.davey
Twitter: @photonstorm

Registered in England and Wales.
Company ID: 8036404
VAT Number: 136 4333 27

**Comment by GGAlanSmithee** [comment]
Yeah, it's a viable option, `lodash` does it for example, just something to be aware of

**Comment by willcode4food** [comment]
Just wanted follow up on this topic and see if we have figured out the next steps.  A lot of great ideas here and I don't want to lose momentum on this aspect of the project.  

From what I understand, perhaps it needs to be decided how Lazer will be structured and what modules need to be shared dependencies and which are exposed to the developer.  When these decisions have been made, I can restructure the imports and exports to make it build accordingly.  Also, I will continue work with lerna to see if it is a good fit.  

Please let know your thoughts.  Thanks!

**Comment by willcode4food** [comment]
Found a monorepo that is much more simple than babel.  I think it could provide some direction on how to structure lazer as a monorepo [Example Monorepo](https://github.com/RickWong/fetch-plus)

**Comment by ivan-kleshnin** [comment]
To support Webpack (and proper bundling in general) `main` entry in `package.json` should lead to source file, not to build...
